src/evaluators/JSXElement.js

```diff
@@ -215,104 +194,113 @@ function evaluateJSXChildren(
-): { attributes: Map<string, Value>, children: ArrayValue | Value | null } {
-  let attributes = new Map();
-  let children = evaluateJSXChildren(astChildren, strictCode, env, realm);
-  let defaultProps = getDefaultProps(elementType, env, realm);
-
-  // defaultProps are a bit like default function arguments
-  // if an actual value exists, it should overwrite the default value
-  if (defaultProps !== null) {
-    for (let [key] of defaultProps.properties) {
-      let defaultPropValue = Get(realm, defaultProps, key);
+): ObjectValue | AbstractObjectValue {
+  let config = Create.ObjectCreate(realm, realm.intrinsics.ObjectPrototype);
+  // start by having key and ref deleted, if they actually exist, they will be added later
+  deleteRefAndKeyFromProps(realm, config);
+  let abstractPropsArgs = [];
+  let containsAbstractSpreadAttribute = false;
+  let spreadAttributesCount = 0;
+  let spreadAttributesWithInitialPropsHintCount = 0;
+  let attributesAssigned = 0;
+  let spreadValue;
+
+  const setConfigProperty = (name: string, value: Value): void => {
+    invariant(config instanceof ObjectValue);
+    Properties.Set(realm, config, name, value, true);
+    attributesAssigned++;
+  };
 
-      if (defaultPropValue instanceof Value) {
-        if (key === "children") {
-          if (children === null) {
-            children = defaultPropValue;
-          }
-        } else {
-          attributes.set(key, defaultPropValue);
-        }
-      }
-    }
-  }
   for (let astAttribute of astAttributes) {
     switch (astAttribute.type) {
       case "JSXAttribute":
         let { name, value } = astAttribute;
 
         invariant(name.type === "JSXIdentifier", `JSX attribute name type not supported: ${astAttribute.type}`);
-        attributes.set(name.name, evaluateJSXValue(((value: any): BabelNodeJSXIdentifier), strictCode, env, realm));
+        setConfigProperty(name.name, evaluateJSXValue(((value: any): BabelNodeJSXIdentifier), strictCode, env, realm));
         break;
       case "JSXSpreadAttribute":
-        let spreadValue = Environment.GetValue(realm, env.evaluate(astAttribute.argument, strictCode));
+        spreadValue = Environment.GetValue(realm, env.evaluate(astAttribute.argument, strictCode));
 
-        if (spreadValue instanceof ObjectValue) {
-          for (let [key, spreadProp] of spreadValue.properties) {
-            if (spreadProp !== undefined && spreadProp.descriptor !== undefined) {
-              let spreadPropValue = spreadProp.descriptor.value;
-
-              if (spreadPropValue instanceof Value) {
-                if (key === "children") {
-                  children = spreadPropValue;
-                } else {
-                  attributes.set(key, spreadPropValue);
-                }
-              }
+        if (spreadValue instanceof ObjectValue && !spreadValue.isPartialObject()) {
+          for (let [spreadPropKey, binding] of spreadValue.properties) {
+            if (binding && binding.descriptor) {
```

Comment:
I think this needs `&& binding.descriptor.enumerable` to fix https://github.com/facebook/prepack/pull/1447#issuecomment-367359080.

Reply:
Great point.